### ISSUE TITLE
Question about create_data.py in the nuSences dataset

### ISSUE THREAD

**Issue by muzi2045** [traveller59/second.pytorch#134]
When tried to using create_data.py to generate .pkl file in the nuSences datasets, there have two command
```
python create_data.py nuscenes_data_prep --data_path=NUSCENES_TRAINVAL_DATASET_ROOT --version="v1.0-trainval" --max_sweeps=9
python create_data.py nuscenes_data_prep --data_path=NUSCENES_TEST_DATASET_ROOT --version="v1.0-test" --max_sweeps=9
```
the first is work in trainval directory, but the second command will occur an error because in the test directory has no infos_train.pkl file.
here is the code:
```
def nuscenes_data_prep(root_path, version, max_sweeps=9):
    nu_ds.create_nuscenes_infos(root_path, version=version, max_sweeps=max_sweeps)
    create_groundtruth_database("NuScenesDataset", root_path, Path(root_path) / "infos_train.pkl")
```

And another thing is training on the nuSences datasets looks like really slow with the car.lite.nu.config(Titan Xp ).

PS: the "data_path" should change to "root_path" in the create_data command.

@traveller59

**Comment by muzi2045** [comment]
there is another thing about infos_train.pkl and infos_val.pkl files generate by create_data.py.
I have download the trainval dataset(part1+part2) and the all test dataset. 
the infos_val.pkl file generated is much larger than infos_train.pkl. As a consequenece, when training into the eval step, an error will occur that some Lidar.bin file can't find in the dataset directory.

**Comment by traveller59** [comment]
@muzi2045 if you want to use nusc evaluation, you MUST download all data because nusc eval function don't support custom data evaluation. otherwise you need to comment evaluation_nusc function (i.e. only use KITTI evaluation).

I will work on nuscenes network training and bug fix these days.

**Comment by traveller59** [comment]
The slow training speed is due to shuffle. I will disable it in all nuscenes config.

**Comment by muzi2045** [comment]
thanks for your reply.
I have modify the shuffle points config to false in train_input_reader and comment the nuse_evaluation funtion.
But training with part1+part2 dataset still looks like the same slow as before.
Maybe this part code in train.py need to modified?
```
dataloader = torch.utils.data.DataLoader(
        dataset,
        batch_size=input_cfg.batch_size,
        shuffle=True,
        num_workers=input_cfg.preprocess.num_workers,
        pin_memory=False,
        collate_fn=merge_second_batch,
        worker_init_fn=_worker_init_fn)
```

**Comment by traveller59** [comment]
- Maybe this part code in train.py need to modified?
- no. 
what's your step_time? the train speed of nuscenes dataset is slower than kitti because too much voxels (10 sweeps) and almost 2x larger detection range area.
my step time is 0.5 with batch size = 6 and 1080Ti.

**Comment by muzi2045** [comment]
```
runtime.step=9950, runtime.steptime=23.31, loss.cls_loss=0.5055, loss.cls_loss_rt=0.5852, loss.loc_loss=0.8417, loss.loc_loss_rt=1.339, loss.loc_elem=[0.01123, 0.0189, 0.3124, 0.05496, 0.0417, 0.07156, 0.1586], loss.cls_pos_rt=0.5192, loss.cls_neg_rt=0.06597, loss.dir_rt=0.6602, rpn_acc=0.9994, pr.prec@10=0.07138, pr.rec@10=0.4882, pr.prec@30=0.6292, pr.rec@30=0.2265, pr.prec@50=0.9094, pr.rec@50=0.109, pr.prec@70=0.9831, pr.rec@70=0.0275, pr.prec@80=1.0, pr.rec@80=0.006216, pr.prec@90=1.0, pr.rec@90=0.0002622, pr.prec@95=1.0, pr.rec@95=8.457e-06, misc.num_vox=154456, misc.num_pos=21, misc.num_neg=124946, misc.num_anchors=125000, misc.lr=0.001053
runtime.step=10000, runtime.steptime=0.6092, loss.cls_loss=0.5058, loss.cls_loss_rt=0.4766, loss.loc_loss=0.8432, loss.loc_loss_rt=0.8381, loss.loc_elem=[0.009259, 0.0139, 0.1382, 0.03474, 0.02926, 0.05298, 0.1408], loss.cls_pos_rt=0.3964, loss.cls_neg_rt=0.08024, loss.dir_rt=0.7014, rpn_acc=0.9994, pr.prec@10=0.07045, pr.rec@10=0.484, pr.prec@30=0.6296, pr.rec@30=0.2231, pr.prec@50=0.9117, pr.rec@50=0.1063, pr.prec@70=0.9852, pr.rec@70=0.02623, pr.prec@80=1.0, pr.rec@80=0.0056, pr.prec@90=1.0, pr.rec@90=0.0002134, pr.prec@95=1.0, pr.rec@95=6.467e-06, misc.num_vox=160000, misc.num_pos=35, misc.num_neg=124913, misc.num_anchors=125000, misc.lr=0.001059
runtime.step=10050, runtime.steptime=0.4527, loss.cls_loss=0.5053, loss.cls_loss_rt=0.4997, loss.loc_loss=0.8364, loss.loc_loss_rt=0.7752, loss.loc_elem=[0.01679, 0.01215, 0.1331, 0.02273, 0.02369, 0.04427, 0.1349], loss.cls_pos_rt=0.4265, loss.cls_neg_rt=0.07322, loss.dir_rt=0.6393, rpn_acc=0.9994, pr.prec@10=0.07069, pr.rec@10=0.4849, pr.prec@30=0.6301, pr.rec@30=0.2235, pr.prec@50=0.9125, pr.rec@50=0.1064, pr.prec@70=0.9843, pr.rec@70=0.02616, pr.prec@80=0.9991, pr.rec@80=0.005606, pr.prec@90=1.0, pr.rec@90=0.0002433, pr.prec@95=1.0, pr.rec@95=5.289e-06, misc.num_vox=160000, misc.num_pos=208, misc.num_neg=124483, misc.num_anchors=125000, misc.lr=0.001066
runtime.step=10100, runtime.steptime=0.6433, loss.cls_loss=0.5089, loss.cls_loss_rt=0.5013, loss.loc_loss=0.8341, loss.loc_loss_rt=0.8802, loss.loc_elem=[0.01357, 0.01518, 0.166, 0.02959, 0.02059, 0.04849, 0.1467], loss.cls_pos_rt=0.4238, loss.cls_neg_rt=0.07756, loss.dir_rt=0.6298, rpn_acc=0.9994, pr.prec@10=0.07055, pr.rec@10=0.4834, pr.prec@30=0.6282, pr.rec@30=0.222, pr.prec@50=0.9137, pr.rec@50=0.1049, pr.prec@70=0.9853, pr.rec@70=0.02551, pr.prec@80=0.9992, pr.rec@80=0.005452, pr.prec@90=1.0, pr.rec@90=0.0002193, pr.prec@95=1.0, pr.rec@95=4.476e-06, misc.num_vox=160000, misc.num_pos=86, misc.num_neg=124772, misc.num_anchors=125000, misc.lr=0.001073
runtime.step=10150, runtime.steptime=0.446, loss.cls_loss=0.5108, loss.cls_loss_rt=0.4573, loss.loc_loss=0.8391, loss.loc_loss_rt=0.6339, loss.loc_elem=[0.005611, 0.01337, 0.1277, 0.03855, 0.03489, 0.05188, 0.04496], loss.cls_pos_rt=0.3062, loss.cls_neg_rt=0.1511, loss.dir_rt=0.5225, rpn_acc=0.9994, pr.prec@10=0.07083, pr.rec@10=0.4799, pr.prec@30=0.627, pr.rec@30=0.2209, pr.prec@50=0.9134, pr.rec@50=0.1037, pr.prec@70=0.9863, pr.rec@70=0.02488, pr.prec@80=0.9993, pr.rec@80=0.005393, pr.prec@90=1.0, pr.rec@90=0.0002483, pr.prec@95=1.0, pr.rec@95=3.88e-06, misc.num_vox=160000, misc.num_pos=27, misc.num_neg=124933, misc.num_anchors=125000, misc.lr=0.00108
runtime.step=10200, runtime.steptime=0.4731, loss.cls_loss=0.5111, loss.cls_loss_rt=0.5579, loss.loc_loss=0.8356, loss.loc_loss_rt=1.157, loss.loc_elem=[0.02037, 0.01148, 0.2843, 0.03015, 0.02495, 0.05002, 0.1571], loss.cls_pos_rt=0.5073, loss.cls_neg_rt=0.0506, loss.dir_rt=0.5882, rpn_acc=0.9994, pr.prec@10=0.07119, pr.rec@10=0.478, pr.prec@30=0.6288, pr.rec@30=0.2203, pr.prec@50=0.9151, pr.rec@50=0.1027, pr.prec@70=0.9868, pr.rec@70=0.02409, pr.prec@80=0.9993, pr.rec@80=0.005105, pr.prec@90=1.0, pr.rec@90=0.0002327, pr.prec@95=1.0, pr.rec@95=3.422e-06, misc.num_vox=160000, misc.num_pos=53, misc.num_neg=124859, misc.num_anchors=125000, misc.lr=0.001087
runtime.step=10250, runtime.steptime=7.232, loss.cls_loss=0.5114, loss.cls_loss_rt=0.4207, loss.loc_loss=0.8364, loss.loc_loss_rt=0.6601, loss.loc_elem=[0.01244, 0.009916, 0.1217, 0.02913, 0.02101, 0.05717, 0.07871], loss.cls_pos_rt=0.3339, loss.cls_neg_rt=0.08681, loss.dir_rt=0.6303, rpn_acc=0.9994, pr.prec@10=0.07103, pr.rec@10=0.4788, pr.prec@30=0.63, pr.rec@30=0.2208, pr.prec@50=0.9146, pr.rec@50=0.1031, pr.prec@70=0.9868, pr.rec@70=0.02424, pr.prec@80=0.9994, pr.rec@80=0.005142, pr.prec@90=1.0, pr.rec@90=0.0002132, pr.prec@95=1.0, pr.rec@95=3.09e-06, misc.num_vox=160000, misc.num_pos=36, misc.num_neg=124909, misc.num_anchors=125000, misc.lr=0.001094
runtime.step=10300, runtime.steptime=0.4649, loss.cls_loss=0.5137, loss.cls_loss_rt=0.5525, loss.loc_loss=0.8384, loss.loc_loss_rt=0.8277, loss.loc_elem=[0.01548, 0.01206, 0.1279, 0.02657, 0.02197, 0.04485, 0.165], loss.cls_pos_rt=0.4723, loss.cls_neg_rt=0.08022, loss.dir_rt=0.6054, rpn_acc=0.9994, pr.prec@10=0.07091, pr.rec@10=0.4764, pr.prec@30=0.6299, pr.rec@30=0.2186, pr.prec@50=0.9144, pr.rec@50=0.1012, pr.prec@70=0.9874, pr.rec@70=0.02352, pr.prec@80=0.9994, pr.rec@80=0.004962, pr.prec@90=1.0, pr.rec@90=0.0002069, pr.prec@95=1.0, pr.rec@95=2.795e-06, misc.num_vox=160000, misc.num_pos=227, misc.num_neg=124418, misc.num_anchors=125000, misc.lr=0.001101
runtime.step=10350, runtime.steptime=0.4574, loss.cls_loss=0.5136, loss.cls_loss_rt=0.5658, loss.loc_loss=0.8377, loss.loc_loss_rt=1.011, loss.loc_elem=[0.0156, 0.01865, 0.2107, 0.02921, 0.03219, 0.04239, 0.1566], loss.cls_pos_rt=0.4833, loss.cls_neg_rt=0.08257, loss.dir_rt=0.6741, rpn_acc=0.9994, pr.prec@10=0.07075, pr.rec@10=0.4768, pr.prec@30=0.6309, pr.rec@30=0.2188, pr.prec@50=0.9149, pr.rec@50=0.1019, pr.prec@70=0.9878, pr.rec@70=0.02398, pr.prec@80=0.9995, pr.rec@80=0.005105, pr.prec@90=1.0, pr.rec@90=0.0002043, pr.prec@95=1.0, pr.rec@95=2.554e-06, misc.num_vox=160000, misc.num_pos=35, misc.num_neg=124910, misc.num_anchors=125000, misc.lr=0.001107
runtime.step=10400, runtime.steptime=5.778, loss.cls_loss=0.5128, loss.cls_loss_rt=0.48, loss.loc_loss=0.8342, loss.loc_loss_rt=0.8668, loss.loc_elem=[0.01403, 0.00886, 0.1854, 0.03316, 0.02491, 0.03614, 0.1309], loss.cls_pos_rt=0.43, loss.cls_neg_rt=0.05004, loss.dir_rt=0.682, rpn_acc=0.9994, pr.prec@10=0.0709, pr.rec@10=0.4781, pr.prec@30=0.629, pr.rec@30=0.2197, pr.prec@50=0.9149, pr.rec@50=0.1021, pr.prec@70=0.9871, pr.rec@70=0.02374, pr.prec@80=0.9986, pr.rec@80=0.005018, pr.prec@90=1.0, pr.rec@90=0.0001874, pr.prec@95=1.0, pr.rec@95=2.343e-06, misc.num_vox=160000, misc.num_pos=55, misc.num_neg=124850, misc.num_anchors=125000, misc.lr=0.001114
runtime.step=10450, runtime.steptime=0.4596, loss.cls_loss=0.5117, loss.cls_loss_rt=0.503, loss.loc_loss=0.8303, loss.loc_loss_rt=0.8378, loss.loc_elem=[0.01483, 0.01349, 0.1437, 0.02217, 0.02681, 0.0417, 0.1563], loss.cls_pos_rt=0.4189, loss.cls_neg_rt=0.08404, loss.dir_rt=0.6522, rpn_acc=0.9994, pr.prec@10=0.07068, pr.rec@10=0.4781, pr.prec@30=0.6287, pr.rec@30=0.2201, pr.prec@50=0.9147, pr.rec@50=0.103, pr.prec@70=0.9868, pr.rec@70=0.02394, pr.prec@80=0.9987, pr.rec@80=0.00503, pr.prec@90=1.0, pr.rec@90=0.0001831, pr.prec@95=1.0, pr.rec@95=4.36e-06, misc.num_vox=160000, misc.num_pos=23, misc.num_neg=124938, misc.num_anchors=125000, misc.lr=0.001121
runtime.step=10500, runtime.steptime=0.4585, loss.cls_loss=0.5102, loss.cls_loss_rt=0.4426, loss.loc_loss=0.8281, loss.loc_loss_rt=0.6635, loss.loc_elem=[0.011, 0.01485, 0.09487, 0.03142, 0.02468, 0.05219, 0.1027], loss.cls_pos_rt=0.3792, loss.cls_neg_rt=0.06335, loss.dir_rt=0.5804, rpn_acc=0.9994, pr.prec@10=0.0707, pr.rec@10=0.4793, pr.prec@30=0.6288, pr.rec@30=0.2208, pr.prec@50=0.9157, pr.rec@50=0.1039, pr.prec@70=0.9875, pr.rec@70=0.02424, pr.prec@80=0.9988, pr.rec@80=0.005126, pr.prec@90=1.0, pr.rec@90=0.0001884, pr.prec@95=1.0, pr.rec@95=1.013e-05, misc.num_vox=160000, misc.num_pos=32, misc.num_neg=124926, misc.num_anchors=125000, misc.lr=0.001128
runtime.step=10550, runtime.steptime=5.74, loss.cls_loss=0.5099, loss.cls_loss_rt=0.532, loss.loc_loss=0.828, loss.loc_loss_rt=0.9141, loss.loc_elem=[0.01355, 0.01369, 0.1785, 0.02197, 0.03103, 0.06204, 0.1363], loss.cls_pos_rt=0.4563, loss.cls_neg_rt=0.07576, loss.dir_rt=0.641, rpn_acc=0.9994, pr.prec@10=0.07046, pr.rec@10=0.4794, pr.prec@30=0.6288, pr.rec@30=0.2211, pr.prec@50=0.9159, pr.rec@50=0.104, pr.prec@70=0.9878, pr.rec@70=0.02429, pr.prec@80=0.9989, pr.rec@80=0.005202, pr.prec@90=1.0, pr.rec@90=0.0001937, pr.prec@95=1.0, pr.rec@95=9.496e-06, misc.num_vox=160000, misc.num_pos=97, misc.num_neg=124750, misc.num_anchors=125000, misc.lr=0.001135
runtime.step=10600, runtime.steptime=0.5183, loss.cls_loss=0.5095, loss.cls_loss_rt=0.5781, loss.loc_loss=0.829, loss.loc_loss_rt=1.163, loss.loc_elem=[0.02013, 0.01314, 0.2919, 0.03929, 0.03625, 0.04434, 0.1365], loss.cls_pos_rt=0.5115, loss.cls_neg_rt=0.06657, loss.dir_rt=0.6089, rpn_acc=0.9994, pr.prec@10=0.07039, pr.rec@10=0.4804, pr.prec@30=0.6291, pr.rec@30=0.2218, pr.prec@50=0.9167, pr.rec@50=0.1043, pr.prec@70=0.9878, pr.rec@70=0.02442, pr.prec@80=0.999, pr.rec@80=0.005271, pr.prec@90=1.0, pr.rec@90=0.0001983, pr.prec@95=1.0, pr.rec@95=1.072e-05, misc.num_vox=160000, misc.num_pos=91, misc.num_neg=124776, misc.num_anchors=125000, misc.lr=0.001142
```
here is the newest log, about the steptime.

**Comment by traveller59** [comment]
preprocess is bottleneck, consider measure time of ```prep_pointcloud``` and ```get_sensor_data``` in NuScenesDataset. In my situation, the np.random.shuffle is bottleneck.
